### PR TITLE
RFC: topology: Add support for HDA backend DAI link configuration

### PR DESCRIPTION
**Pull request by sofci**
I'm still not familiar with m4 now to directly revise existing DAI_CONFIG(type, idx...) to add support for HDA BE links, and it seems current HW configuration section is more suitable to SSP links but not to HD-A links. So I define a new macro HDA_DAI_CONFIG(idx, ...) for users to define HDA BE link configurations.

Later, when we're more confident about what to configure for HDA links, we can revise the implementation. Now this can provide a fast way to add HDA BE links in machine topology file. 

The patch was verified with nocodec machine driver on Leaf Hill.

### CODE REVIEW

File: topology/m4/dai.m4 [hunk 1]
```diff
@@ -160,6 +160,29 @@ define(`DAI_CONFIG',
 `	]'
 `}')
 
+dnl HDA_DAI_CONFIG(idx, link_id, name)
```
Comment (lgirdwood) [inline topology/m4/dai.m4]: Best to put this in hda.m4

**Comment by mengdonglin** [comment]
I'll revise as Liam commented in v2. Thanks for review.

File: topology/sof-apl-hda.m4 [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+include(`sof/tokens.m4')
+
+# Include Apollolake DSP configuration
+include(`platform/intel/bxt.m4')
+
+#
+# Define the pipelines
+#
+# PCM0 ----> volume -----> iDisp1
+# PCM1 ----> Volume <----- iDisp2
```
Comment (plbossart) [inline topology/sof-apl-hda.m4]: fix arrow direction

File: topology/sof-apl-hda.m4 [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+dnl DAI_ADD(pipeline,
+dnl     pipe id, dai type, dai_index, dai_be,
+dnl     buffer, periods, format,
+dnl     frames, deadline, priority, core)
+
+# playback DAI is iDisp1 using 2 periods
+# Buffers use s16le format, with 48 frame per 1000us on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-playback.m4,
+	1, HDA, 0, iDisp1,
+	PIPELINE_SOURCE_1, 2, s16le,
```
Comment (plbossart) [inline topology/sof-apl-hda.m4]: This is a problem: we are restricting the output capabilities regardless of what the receiver can do - as exposed by the ELD. We should have something that essentially does not constrain the output by default (it'd be fine to allow for constraints to be added for debug, but by default we should let the hardware adjust).

File: topology/Makefile.am [hunk 4]
```diff
@@ -19,6 +19,7 @@ DEPS = \
 MACHINES = \
 	sof-cht-nocodec.tplg \
 	sof-cht-max98090.tplg \
+	sof-apl-hda.tplg \
```
Comment (plbossart) [inline topology/Makefile.am]: sof-apl-hdmi ?
This is not related to HDaudio, if you wanted to deal with an HDaudio codec you'd need additional PCMs and outputs.

File: topology/sof-apl-hda.m4 (hunk 2 above)
Comment (sofci) [reply topology/sof-apl-hda.m4]: Okay, I will fix this in v3. I also need to revise the PCM name for HDMI from Portx to HDMIx, so that the generic DSP HDA driver can find the HDMI links by checking the link name and do proper inItialization for HDMI.

File: topology/sof-apl-hda.m4 (hunk 3 above)
Comment (sofci) [reply topology/sof-apl-hda.m4]: I agree with you, these constrains should only be for debug, but we need a better solution to remove the constrains later.

HDMI/Displayport can support a wide range of compositions of sample rates, formats and channels. It can also support non-PCM format e.g. HD-DTS that we've never made it work. A long list of all possible configurations would be ugly and may not work.

File: topology/Makefile.am (hunk 4 above)
Comment (sofci) [reply topology/Makefile.am]: I'm using sof-apl-hda because I use it to make general DSP HDA driver (skl_hda_generic) work with SOF.  After HDMI works, I'll add analog info into this m4 file.  

Thanks
Mengdong
